### PR TITLE
Implemented Disable Gardener provided static admin kubeconfig

### LINKED ISSUE #1186: Allow the user to disable/enable the admin kubeconfig
**What would you like to be added**:
The static admin kubeconfig for shoot clusters can now be disabled by setting ~~`.spec.kubernetes.enableAdminKubeconfig=false`~~ in the specification of the `Shoot` resource. The respective `<shoot-name>.kubeconfig` secret in the project namespace in the garden cluster will be deleted.
There should be an option on the dashboard to do this for new and already existing clusters. Ref https://github.com/gardener/gardener/pull/5649

UPDATE: the filed has changed to `.spec.kubernetes.enableStaticTokenKubeconfig`

**Why is this needed**:
https://github.com/gardener/gardener/issues/3138

### PR DESCRIPTION
**Pull request by grolu**
**What this PR does / why we need it**:
Support to disable admin kubeconfig during cluster creation and on shoot details page

Screen Shot 2022-08-16 at 14 00 01
Screen Shot 2022-08-16 at 14 00 15
Screen Shot 2022-07-06 at 14 30 58

**Which issue(s) this PR fixes**:
Fixes #1186

**Special notes for your reviewer**:

**Release note**:

```feature user
Static token kubeconfig cluster access can be disabled for new and existing clusters
```

### CODE REVIEW

**Comment by rfranzke** [comment]
Add a link to https://github.com/gardener/gardener/blob/master/docs/usage/shoot_access.md

File: backend/test/acceptance/__snapshots__/api.shoots.spec.js.snap [hunk 1]
```diff
@@ -981,7 +1037,7 @@ users:
 }
 `;
 
-exports[`api shoots should return shoot seed info when no fallback is needed 1`] = `
+exports[`api shoots should return shoot seed info when need to fallback to old monitoring secret 1`] = `
```
Comment (holgerkoser) [inline backend/test/acceptance/__snapshots__/api.shoots.spec.js.snap:1040]: Why is this changed. Has the order changed? Can we revert it?

File: frontend/src/components/AdminKubeconfig.vue [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      v-model="enableStaticTokenKubeconfig"
+      label="Enable static token kubeconfig"
+      color="primary"
+      @change="onChange"
+      hide-details
+    ></v-checkbox>
+    <span class="text-caption">It is <span class="font-weight-bold">not</span> the recommended method to access the shoot cluster as the static token <code>kubeconfig</code> has some security flaws associated with it.</span>
+    <p>
+      <external-link class="text-caption" url="https://github.com/gardener/gardener/blob/master/docs/usage/shoot_access.md">More information</external-link>
+    </p>
```
Comment (holgerkoser) [inline frontend/src/components/AdminKubeconfig.vue:19]:
```suggestion
    <div>
      <external-link class="text-caption" url="https://github.com/gardener/gardener/blob/master/docs/usage/shoot_access.md">More information</external-link>
    </div>
```

File: frontend/src/components/AdminKubeconfig.vue [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  methods: {
+    onChange (value) {
+      this.$emit('input', value)
+    }
+  },
+  watch: {
+    value: function (value) {
+      this.enableStaticTokenKubeconfig = value
+    }
+  }
```
Comment (holgerkoser) [inline frontend/src/components/AdminKubeconfig.vue:50]:
```suggestion
  props: {
    value: {
      type: Boolean,
      required: true
    }
  },
  computed: {
    enableStaticTokenKubeconfig: {
      get () {
        return this.value
      },
      set (val) {
        this.$emit('input', val)
      }
    }
  }
```

Comment (holgerkoser) [reply frontend/src/components/AdminKubeconfig.vue:50]: Or is there a reason to only handle the change event?

File: frontend/src/components/AdminKubeconfig.vue [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+SPDX-License-Identifier: Apache-2.0
+-->
+
+<template>
+  <div>
+    <v-checkbox
+      v-model="enableStaticTokenKubeconfig"
+      label="Enable static token kubeconfig"
+      color="primary"
+      @change="onChange"
```
Comment (holgerkoser) [inline frontend/src/components/AdminKubeconfig.vue:13]:
```suggestion
```

File: backend/lib/services/shoots.js [hunk 5]
```diff
@@ -82,6 +82,17 @@ exports.replaceVersion = async function ({ user, namespace, name, body }) {
   return client['core.gardener.cloud'].shoots.jsonPatch(namespace, name, patchOperations)
 }
 
+exports.replaceEnableStaticTokenKubeconfig = async function ({ user, namespace, name, body }) {
+  const client = user.client
+  const enableStaticTokenKubeconfig = body.enableStaticTokenKubeconfig
```
[the remark this answers is not in the document]
Comment (grolu) [reply backend/lib/services/shoots.js:87]: done

File: frontend/src/components/AdminKubeconfig.vue [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+<template>
+  <div>
+    <v-checkbox
+      v-model="enableStaticTokenKubeconfig"
+      label="Enable static token kubeconfig"
```
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/components/AdminKubeconfig.vue:11]: I checked other checkbox labels, we always have them like sentences, e.g. `This ${purpose} cluster does not need a hibernation schedule`

File: frontend/src/components/AdminKubeconfig.vue [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+      <external-link class="text-caption" url="https://github.com/gardener/gardener/blob/master/docs/usage/shoot_access.md">More information</external-link>
+    </p>
+  </div>
+</template>
+<script>
+
+import ExternalLink from '@/components/ExternalLink.vue'
+
+export default {
+  name: 'AdminKubeconfig',
```
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/components/AdminKubeconfig.vue:27]: Yes, You are right I guess. I tried to align with other components, like `Purpose` should rather be called `PurposeSelect`, right? That's why I did not add `Switch` to the component name. And it even gets better we have the `SelectSecret` component which should rather be called `SecretSelect`, right.?I will rename as you suggested, we do not have a clean / consistent naming anyway.

File: frontend/src/components/AdminKubeconfigConfiguration.vue [hunk 8]
```diff
@@ -0,0 +1,73 @@
+<!--
+SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+<template>
+  <action-button-dialog
```
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/components/AdminKubeconfigConfiguration.vue:8]: okay, I added the confirmation prop

File: frontend/src/components/ShootDetails/ShootAccessCard.vue [hunk 9]
```diff
@@ -27,7 +27,7 @@ SPDX-License-Identifier: Apache-2.0
       >
     </terminal-list-tile>
 
-    <v-divider v-if="isTerminalTileVisible && (isTerminalShortcutsTileVisible || isDashboardTileVisible || isCredentialsTileVisible || isKubeconfigTileVisible || isGardenctlTileVisible)" inset></v-divider>
+    <v-divider v-if="isTerminalTileVisible" inset></v-divider>
```
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/components/ShootDetails/ShootAccessCard.vue:30]: Yes, we should discuss this

File: frontend/src/components/ShootDetails/ShootAccessCard.vue [hunk 10]
```diff
@@ -88,20 +88,22 @@ SPDX-License-Identifier: Apache-2.0
...
 
-    <v-list-item v-if="isKubeconfigTileVisible">
+    <v-list-item>
       <v-list-item-icon>
         <v-icon color="primary">mdi-file</v-icon>
       </v-list-item-icon>
       <v-list-item-content>
         <v-list-item-title>Kubeconfig</v-list-item-title>
+        <v-list-item-subtitle v-if="!shootEnableStaticTokenKubeconfig">Static token kubeconfig is disabled for this cluster</v-list-item-subtitle>
+        <v-list-item-subtitle v-else-if="!isKubeconfigAvailable">Static token kubeconfig currently not available</v-list-item-subtitle>
```
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/components/ShootDetails/ShootAccessCard.vue:104]: I cannot confirm this - it works for me. have you tried restarting your local backend server? Recently, I had similar issues with broken watch connections in the backend, resulting in not reactive frontend UI (locally)

File: frontend/src/components/AdminKubeconfig.vue (hunk 3 above)
Comment (grolu) [reply frontend/src/components/AdminKubeconfig.vue]: No there is no reason, I changed it

File: backend/test/acceptance/__snapshots__/api.shoots.spec.js.snap (hunk 1 above)
Comment (holgerkoser) [reply backend/test/acceptance/__snapshots__/api.shoots.spec.js.snap:1040]: It seems that the initial order of the snapshots https://github.com/gardener/dashboard/pull/1178/files#diff-7e96d3311ad28585b977e916709b718e22e93aaac7e42ad86b450697722355f9 was wrong. Only explanation I have the entries have been added manually.

File: frontend/src/components/ShootDetails/ShootAccessCard.vue (hunk 10 above)
[the remark this answers is not in the document]
Comment (petersutter) [reply frontend/src/components/ShootDetails/ShootAccessCard.vue:104]: i ran `yarn` again and now it works

File: backend/test/acceptance/__snapshots__/api.shoots.spec.js.snap (hunk 1 above)
Comment (grolu) [reply backend/test/acceptance/__snapshots__/api.shoots.spec.js.snap:1040]: so I guess we go for the generated order, right?

File: frontend/src/components/AdminKubeconfig.vue [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+<template>
+  <div>
+    <v-checkbox
+      v-model="enableStaticTokenKubeconfig"
+      label="Enable static token kubeconfig"
+      color="primary"
+      @change="onChange"
+      hide-details
+    ></v-checkbox>
+    <span class="text-caption">It is <span class="font-weight-bold">not</span> the recommended method to access the shoot cluster as the static token <code>kubeconfig</code> has some security flaws associated with it.</span>
```
[the remark this answers is not in the document]
Comment (petersutter) [reply frontend/src/components/AdminKubeconfig.vue]:
```js
<div class="text-caption mt-1">Not recommended. There are more <external-link class="text-caption" url="https://github.com/gardener/gardener/blob/master/docs/usage/shoot_access.md">secure alternatives</external-link> to access the Shoot cluster.</div>
```

Screenshot 2022-07-27 at 16 57 11

@holgerkoser WDYT?

[the remark this answers is not in the document]
Comment (petersutter) [reply frontend/src/components/AdminKubeconfig.vue]: suggestion from @holgerkoser  👍 
Screenshot 2022-07-27 at 17 11 44

File: frontend/src/components/StaticTokenKubeconfigSwitch.vue [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+
+<template>
+  <div>
+    <v-checkbox
+      v-model="enableStaticTokenKubeconfig"
+      label="Enable static token kubeconfig"
+      color="primary"
+      hide-details
+    ></v-checkbox>
+    <div class="text-caption mt-1">It is <strong>not</strong> the recommended method to access the shoot cluster as the static token <code>kubeconfig</code> has some security flaws associated with it.</div>
```
Comment (petersutter) [inline frontend/src/components/StaticTokenKubeconfigSwitch.vue:15]:
```suggestion
      label="Enable static token kubeconfig (not recommended)"
      color="primary"
      hide-details
    ></v-checkbox>
    <div class="text-caption mt-1">There are more <external-link class="text-caption" url="https://github.com/gardener/gardener/blob/master/docs/usage/shoot_access.md">secure alternatives</external-link> to access the Shoot cluster.</div>
```

**Comment by rfranzke** [comment]
Nice!

File: backend/test/acceptance/__snapshots__/api.shoots.spec.js.snap (hunk 1 above)
Comment (holgerkoser) [reply backend/test/acceptance/__snapshots__/api.shoots.spec.js.snap:1040]: yes

File: frontend/src/components/ShootDetails/ShootAccessCard.vue (hunk 9 above)
[the remark this answers is not in the document]
Comment (holgerkoser) [reply frontend/src/components/ShootDetails/ShootAccessCard.vue:30]: 🤔

File: frontend/src/components/StaticTokenKubeconfigConfiguration.vue [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+
+<script>
+import ActionButtonDialog from '@/components/dialogs/ActionButtonDialog'
+import StaticTokenKubeconfigSwitch from '@/components/StaticTokenKubeconfigSwitch'
+import { updateShootEnableStaticTokenKubeconfig } from '@/utils/api'
+import { errorDetailsFromError } from '@/utils/error'
+import shootItem from '@/mixins/shootItem'
+
+export default {
+  name: 'StaticTokenKubeconfigConfiguration',
```
Comment (holgerkoser) [inline frontend/src/components/StaticTokenKubeconfigConfiguration.vue:31]:
```suggestion
  name: 'static-token-kubeconfig-configuration',
```

File: frontend/src/components/StaticTokenKubeconfigConfiguration.vue [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+      await this.reset()
+      const confirmed = await this.$refs.actionDialog.waitForDialogClosed()
+      if (confirmed) {
+        await this.updateConfiguration()
+      }
+    },
+    async updateConfiguration () {
+      try {
+        await updateShootEnableStaticTokenKubeconfig({
+          namespace: this.shootNamespace,
```
Comment (holgerkoser) [inline frontend/src/components/StaticTokenKubeconfigConfiguration.vue:55]: I am not sure what it means from a performance and memory point of view to include a large mixin. The only thing we use is the `name` and `namespace` from the shootItem?

File: frontend/src/components/StaticTokenKubeconfigSwitch.vue [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+      <external-link class="text-caption" url="https://github.com/gardener/gardener/blob/master/docs/usage/shoot_access.md">More information</external-link>
+    </div>
+  </div>
+</template>
+<script>
+
+import ExternalLink from '@/components/ExternalLink.vue'
+
+export default {
+  name: 'StaticTokenKubeconfigSwitch',
```
Comment (holgerkoser) [inline frontend/src/components/StaticTokenKubeconfigSwitch.vue:26]:
```suggestion
  name: 'static-token-kubeconfig-switch',
```

File: frontend/src/components/StaticTokenKubeconfigConfiguration.vue [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+        })
+      } catch (err) {
+        const errorMessage = 'Could not update static kubeconfig flag'
+        const errorDetails = errorDetailsFromError(err)
+        const detailedErrorMessage = errorDetails.detailedMessage
+        this.$refs.actionDialog.setError({ errorMessage, detailedErrorMessage })
+        console.error(this.errorMessage, errorDetails.errorCode, errorDetails.detailedMessage, err)
+      }
+    },
+    async reset () {
```
Comment (holgerkoser) [inline frontend/src/components/StaticTokenKubeconfigConfiguration.vue:69]:
```suggestion
    reset () {
```

File: frontend/src/components/StaticTokenKubeconfigConfiguration.vue [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+    shootItem
+  ],
+  data () {
+    return {
+      enableStaticTokenKubeconfig: undefined
+    }
+  },
+  methods: {
+    async onConfigurationDialogOpened () {
+      await this.reset()
```
Comment (holgerkoser) [inline frontend/src/components/StaticTokenKubeconfigConfiguration.vue:46]:
```suggestion
      this.reset()
```

File: frontend/src/components/StaticTokenKubeconfigConfiguration.vue [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+  components: {
+    ActionButtonDialog,
+    StaticTokenKubeconfigSwitch
+  },
+  mixins: [
+    shootItem
+  ],
+  data () {
+    return {
+      enableStaticTokenKubeconfig: undefined
```
Comment (holgerkoser) [inline frontend/src/components/StaticTokenKubeconfigConfiguration.vue:41]:
```suggestion
      enableStaticTokenKubeconfig: this.shootEnableStaticTokenKubeconfig
```

File: frontend/src/views/NewShoot.vue [hunk 19]
```diff
@@ -418,9 +419,10 @@ export default {
 
       const name = get(shootResource, 'metadata.name')
       const kubernetesVersion = get(shootResource, 'spec.kubernetes.version')
+      const enableStaticTokenKubeconfig = get(shootResource, 'spec.kubernetes.enableStaticTokenKubeconfig')
       const purpose = get(shootResource, 'spec.purpose')
       this.purpose = purpose
-      await this.$refs.clusterDetails.setDetailsData({ name, kubernetesVersion, purpose, secret, cloudProfileName, updateK8sMaintenance: k8sUpdates })
+      await this.$refs.clusterDetails.setDetailsData({ name, kubernetesVersion, purpose, secret, cloudProfileName, updateK8sMaintenance: k8sUpdates, enableStaticTokenKubeconfig })
```
Comment (holgerkoser) [inline frontend/src/views/NewShoot.vue:425]:
```suggestion
      await this.$refs.clusterDetails.setDetailsData({ 
        name, 
        kubernetesVersion, 
        purpose, 
        secret, 
        cloudProfileName, 
        updateK8sMaintenance: k8sUpdates, 
        enableStaticTokenKubeconfig 
      })
```

File: frontend/src/views/NewShoot.vue [hunk 20]
```diff
@@ -310,9 +310,10 @@ export default {
         this.$refs.accessRestrictions.applyTo(shootResource)
       }
 
-      const { name, kubernetesVersion, purpose } = this.$refs.clusterDetails.getDetailsData()
+      const { name, kubernetesVersion, purpose, enableStaticTokenKubeconfig } = this.$refs.clusterDetails.getDetailsData()
```
Comment (holgerkoser) [inline frontend/src/views/NewShoot.vue:313]:
```suggestion
      const { 
        name, 
        kubernetesVersion, 
        purpose, 
        enableStaticTokenKubeconfig 
      } = this.$refs.clusterDetails.getDetailsData()
```

File: frontend/src/components/ShootDetails/ShootAccessCard.vue (hunk 9 above)
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/components/ShootDetails/ShootAccessCard.vue:30]: Have a look at the code again - the kubeconfig tile is now _always_ visible, so the `isKubeconfigTileVisible` check has been removed. Therefore I removed all parts of the conditions that evaluate `true` if `isKubeconfigTileVisible` is `true`. Does this make sense?

File: frontend/src/components/StaticTokenKubeconfigConfiguration.vue (hunk 13 above)
Comment (grolu) [reply frontend/src/components/StaticTokenKubeconfigConfiguration.vue:31]: >🤔 The idea was to show the divider if the previous tile is visible and any of the following tiles is visible. Why is the check for the following tiles not necessary anymore?

Why would we need it if kubeconfig tile is always visible

File: frontend/src/components/StaticTokenKubeconfigConfiguration.vue (hunk 14 above)
Comment (grolu) [reply frontend/src/components/StaticTokenKubeconfigConfiguration.vue:55]: IDK either, I also already thought about splitting the mixin up into separate parts as it grew a lot over time. In this case we also use `shootEnableStaticTokenKubeconfig`

File: frontend/src/components/ShootDetails/ShootAccessCard.vue (hunk 9 above)
[the remark this answers is not in the document]
Comment (grolu) [reply frontend/src/components/ShootDetails/ShootAccessCard.vue]: Show kubeconfig tile only if kubeconfig available OR user can patch shoots

File: frontend/src/components/StaticTokenKubeconfigSwitch.vue [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+<template>
+  <div>
+    <v-checkbox
+      v-model="enableStaticTokenKubeconfig"
+      label="Enable static token kubeconfig"
```
Comment (petersutter) [inline frontend/src/components/StaticTokenKubeconfigSwitch.vue:11]:
```suggestion
      label="Enable static token kubeconfig (not recommended)"
```
strange, I thought @holgerkoser  and I already made this suggestion, but somehow it got lost

Comment (grolu) [reply frontend/src/components/StaticTokenKubeconfigSwitch.vue:11]: okay

**Review by petersutter — APPROVED** [review]
/lgtm
